src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala

```diff
@@ -64,10 +66,15 @@ private[health] class HealthCheckActor(
       app.version,
       healthCheck
     )
-    val activeInstanceIds: Set[Instance.Id] = instanceTracker.specInstancesLaunchedSync(app.id).map(_.instanceId)(collection.breakOut)
-    // The Map built with filterKeys wraps the original map and contains a reference to activeInstanceIds.
-    // Therefore we materialize it into a new map.
-    healthByInstanceId = healthByInstanceId.filterKeys(activeInstanceIds).iterator.toMap
+    instanceTracker.instancesBySpec().map(
+      _.specInstances(app.id).filter(_.isLaunched).map(_.instanceId)(collection.breakOut)
+    ).onComplete { r =>
```

Comment:
Not really a fan of this style, instead, consider using FSM and stashing or specialized "Finished" where the logic is in the actor, not in the message itself. While I'd prefer FSM for this, it could probably be done well using become and stashing only as a smaller change.